### PR TITLE
Feature/radical-nitrous-injection

### PR DESCRIPTION
**Pull request by sarahsturgeon**
No errors found as of posting this, but will conduct further testing

The goal of this PR is to make URS less slow at the cost of any readability. We just need it to function as-is with a lower performance impact.

### CODE REVIEW

File: lua/ulx/modules/urs_server.lua [hunk 1]
```diff
@@ -1,193 +1,268 @@
...
 function URS.CheckRestrictedTool(ply, tr, tool)
-	if URS.Check( ply, "tool", tool ) == false then return false end
-    if not IsValid( ply ) then return end
-	if URS.cfg.echoSpawns:GetBool() and tool != "inflator" then
-		ulx.logSpawn( ply:Nick().."<".. ply:SteamID() .."> used the tool ".. tool .." on ".. tr.Entity:GetModel() )
-	end
+    if Check( ply, "tool", tool ) == false then return false end
+
+    if not echoSpawns then return end
+    if ignoredTools[tool] then return end
```
Comment (sarahsturgeon) [inline lua/ulx/modules/urs_server.lua:217]:
```suggestion
    if rawget( ignoredTools, tool ) == true then return end
```

File: lua/ulx/xgui/settings/cl_urs_gui.lua [hunk 2]
```diff
@@ -1 +1,183 @@
...
xgui.hookEvent( "URSLoadouts", "process", URSLoadoutsProcess )

xgui.addSettingModule( "URS", urs.back, "icon16/shield.png" )
\ No newline at end of file
+xgui.prepareDataType( "URSRestrictions" )
+xgui.prepareDataType( "URSLimits" )
+
+urs = {}
+urs.removers = {}
+urs.weapons2 = {"weapon_para","weapon_crowbar","weapon_stunstick","weapon_physcannon","weapon_physgun","weapon_pistol","weapon_357","weapon_smg1","weapon_ar2","weapon_shotgun","weapon_crossbow",  "weapon_frag", "weapon_rpg", "weapon_slam", "weapon_bugbait", "item_ml_grenade", "item_ar2_grenade", "item_ammo_ar2_altfire", "gmod_camera", "gmod_tool"}
```
Comment (sarahsturgeon) [inline lua/ulx/xgui/settings/cl_urs_gui.lua]:
```suggestion
```
I think this can be removed?

File: lua/ulx/modules/urs_server.lua [hunk 3]
```diff
@@ -1,193 +1,268 @@
...
+    for _, ply in ipairs( player.GetAll() ) do
+        ply.URS_CacheCheck = nil
+    end
+end
+
+local echoSpawns
+local overwriteSbox
+local weaponPickups
+
+hook.Add( "Initialize", "URSConfLoad", function()
```
Comment (sarahsturgeon) [inline lua/ulx/modules/urs_server.lua:45]: Will this work on this hook?

File: lua/ulx/modules/urs_server.lua [hunk 4]
```diff
@@ -1,193 +1,268 @@
...
+-- Return cached value for this check
+function URS.CachedCheck(ply, restrictionType, what)
+    ply.URS_CacheCheck = ply.URS_CacheCheck or {}
+
+    local existing = ply.URS_CacheCheck
+    local restrictionTypes = rawget( existing, restrictionType ) 
+
+    if not restrictionTypes then return end
+
+    -- true, false, or nil
```
Comment (sarahsturgeon) [inline lua/ulx/modules/urs_server.lua:75]: TODO: Should be 100% sure of this

File: lua/ulx/modules/urs_server.lua [hunk 5]
```diff
@@ -1,193 +1,268 @@
...
+            return cacheCheck( ply, restrictionType, what, false )
+        end
+    end
+
+    local allRestrictions = rawget( ursRestrictions, "all" )
+    local allTypeRestrictions = allRestrictions and rawget( allRestrictions, restrictionType )
+    local hasGroup = allTypeRestrictions and rawget( allTypeRestrictions, group )
+
+    if hasGroup then
+        ULib.tsayError(ply, "Your rank is restricted from all ".. restrictionTypePlural)
```
Comment (sarahsturgeon) [inline lua/ulx/modules/urs_server.lua]:
```suggestion
        ULib.tsayError( ply, "Your rank is restricted from all ".. restrictionTypePlural )
```

Comment (sarahsturgeon) [reply lua/ulx/modules/urs_server.lua]: Also TODO: Should probably localize `tsayError`

File: lua/ulx/modules/urs_server.lua [hunk 6]
```diff
@@ -1,193 +1,268 @@
...
-			if ply:GetCount(type.."s") >= URS.limits[type][group] then 
-				ply:LimitHit( type .."s" )
-				return false 
-			end 
-		end 
-		if URS.cfg.overwriteSbox:GetBool() then 
-			return true -- Overwrite sbox limit (ours is greater) 
-		end 
-	end 
+    rawset( restrictionTypes, what, result )
```
Comment (wrefgtzweve) [inline lua/ulx/modules/urs_server.lua:90]: [urs] addons/urs/lua/ulx/modules/urs_server.lua:92: bad argument #1 to 'rawset' (table expected, got nil)
  1. rawset - [C]:-1
   2. Check - addons/urs/lua/ulx/modules/urs_server.lua:92
    3. fn - addons/urs/lua/ulx/modules/urs_server.lua:239
     4. Call - lua/ulib/shared/hook.lua:109
      5. unknown - gamemodes/sandbox/gamemode/commands.lua:845
       6. unknown - lua/includes/modules/concommand.lua:54

       when spawning in a restricted weapon, only happens once

Comment (wrefgtzweve) [reply lua/ulx/modules/urs_server.lua:90]: the print only prints once ""weapon_pistol" is a restricted swep from your rank." and doesn't print after trying to spawn anymore

Comment (wrefgtzweve) [reply lua/ulx/modules/urs_server.lua:90]: Fixed by returning early at `if not restrictionTypes then`

File: lua/ulx/modules/urs_server.lua [hunk 7]
```diff
@@ -1,193 +1,268 @@
...
+    if restriction then
+        local hasGroup = rawget( restriction, group )
+        local hasPlayer = rawget( restriction, ply:SteamID() )
+
+        if rawget( restriction, "*" ) then
+            if not ( hasGroup or hasPlayer ) then
+                PrintRestricted( ply, restrictionType, what )
+                return cacheCheck( ply, restrictionType, what, false )
+            end
+        elseif hasGroup or hasPlayer then
```
Comment (wrefgtzweve) [inline lua/ulx/modules/urs_server.lua:128]: Only triggers twice for some reason, any clicks after that will prevent the print from showing up. Spawning a swep will still be blocked though

Comment (wrefgtzweve) [reply lua/ulx/modules/urs_server.lua:128]: Issue found, cached result doesn't print

File: lua/ulx/modules/urs_server.lua (hunk 3 above)
Comment (wrefgtzweve) [reply lua/ulx/modules/urs_server.lua:45]: Works

File: lua/ulx/modules/urs_server.lua (hunk 5 above)
Comment (wrefgtzweve) [reply lua/ulx/modules/urs_server.lua]: Localized it

File: lua/ulx/xgui/settings/cl_urs_gui.lua (hunk 2 above)
Comment (wrefgtzweve) [reply lua/ulx/xgui/settings/cl_urs_gui.lua]: urs.weapons2 can be removed
image

Comment (wrefgtzweve) [reply lua/ulx/xgui/settings/cl_urs_gui.lua]: Removed

**Comment by wrefgtzweve** [comment]
Limits seem not to be working right now, looking into fixing them

Edit:

Found the error, a typo:

```   
local limitsHasType = rawget( ursTypeLimits, restrictionType )
if not limitHasType then return end
```

**Comment by wrefgtzweve** [comment]
image
A lot of these functions are in tables, wouldn't it be better to localize all of these?, especially if they only get used once

EDIT:
I localized all the hooks

File: lua/ulx/modules/urs_server.lua (hunk 7 above)
Comment (wrefgtzweve) [reply lua/ulx/modules/urs_server.lua:128]: Resolved

File: lua/ulx/modules/urs_server.lua [hunk 8]
```diff
@@ -1,193 +1,267 @@
...
+
+    local ursLimits = rawget( URS, "limits" )
+    local typeLimits = rawget( ursLimits, restrictionType )
+    if not typeLimits then return end
+
+    local playerTypeLimit = rawget( typeLimits, ply:SteamID() )
+    local groupTypeLimit = rawget( typeLimits, group )
+    if not playerTypeLimit and not groupTypeLimit then return end
+
+    -- TODO: Should this be an elseif? Shouldn't both cases be possible?
```
Comment (wrefgtzweve) [inline lua/ulx/modules/urs_server.lua:156]: You can only return once, if both cases can exist playerlimit should go first. That's how regular URS does it too.

File: lua/ulx/modules/urs_server.lua [hunk 9]
```diff
@@ -1,193 +1,276 @@
...
+            return false
+        end
+    end
+    if overwriteSbox then
+        return true -- Overwrite sbox limit (ours is greater)
+    end
+end
+local Check = URS.Check
+
+timer.Simple( 0.1, function()
```
Comment (wrefgtzweve) [inline lua/ulx/modules/urs_server.lua:174]: I feel like this can be optimized a lot, maybe not too needed but it's bad in its current state

**Pull request merged**